### PR TITLE
SourceRef in expressions

### PR DESCRIPTION
**Pull request by gzanitti**
This PR solves issue https://github.com/powdr-labs/powdr/issues/1293.

Includes PRs that modify some Expression in order to match fields. It is recommended to review/merge those first: 
- #1351 
- #1352
- #1353
- #1354
- #1355

All comments are welcome :)

### CODE REVIEW

**Comment by lvella** [comment]
Let's merge the dependency PRs before undrafting this one.

File: ast/src/parsed/mod.rs [hunk 1]
```diff
@@ -332,24 +332,119 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+    IndexAccess(SourceRef, IndexAccess<Self>),
+    FunctionCall(SourceRef, FunctionCall<Self>),
+    FreeInput(SourceRef, Box<Self>),
+    MatchExpression(SourceRef, MatchExpression<Self>),
+    IfExpression(SourceRef, IfExpression<Self>),
+    BlockExpression(SourceRef, BlockExpression<Self>),
+}
+
+// Comparison function for expressions that ignore source information.
+impl<Ref: PartialEq> PartialEq for Expression<Ref> {
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:356]: Do you know where this is needed?

File: ast/src/parsed/mod.rs [hunk 2]
```diff
@@ -332,24 +332,119 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+            ArrayLiteral(src, _) => *src = new_src,
+            BinaryOperation(src, _) => *src = new_src,
+            UnaryOperation(src, _) => *src = new_src,
+            IndexAccess(src, _) => *src = new_src,
+            FunctionCall(src, _) => *src = new_src,
+            FreeInput(src, _) => *src = new_src,
+            MatchExpression(src, _) => *src = new_src,
+            IfExpression(src, _) => *src = new_src,
+            BlockExpression(src, _) => *src = new_src,
+        }
```
Comment (chriseth) [inline ast/src/parsed/mod.rs]: *self.source_reference_mut() = new_src;

(actually implement that in the trait)

File: ast/src/parsed/mod.rs [hunk 3]
```diff
@@ -386,17 +481,25 @@ pub struct BinaryOperation<E = Expression<NamespacedPolynomialReference>> {
...
+            SourceRef::unknown(),
+            BinaryOperation {
+                left: Box::new(left),
+                op,
+                right: Box::new(right),
+            },
+        )
+    }
+
+    pub fn with_source_reference(&mut self, s: SourceRef) -> &mut Self {
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:480]: Shouldn't this take and return by value?

File: ast/src/parsed/mod.rs (hunk 1 above)
Comment (gzanitti) [reply ast/src/parsed/mod.rs:356]: I remember this particular case:

https://github.com/powdr-labs/powdr/blob/e2310b1b66a4c85fddee5d8d954a170f42e8b4b8/linker/src/lib.rs#L39

But also many matches against numbers started to fail in tests.

File: parser/src/powdr.lalrpop [hunk 4]
```diff
@@ -450,8 +450,8 @@ BoxedExpression: Box<Expression> = {
 }
 
 LambdaExpression: Box<Expression> = {
-    <kind:FunctionKind> "||" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(LambdaExpression{kind, params: vec![], body})),
-    <kind:FunctionKind> "|" <params:ParameterList> "|" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(LambdaExpression{kind, params, body})),
+    <start:@L> <kind:FunctionKind> "||" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(ctx.source_ref(start), LambdaExpression{kind, params: vec![], body})),
```
[the remark this answers is not in the document]
Comment (gzanitti) [reply parser/src/powdr.lalrpop:453]: > Could be `LambdaExpression{kind, praams: vec![], body}.into().with_source_ref(ctx.source_ref(start))`

I tried to implement it that way, but Rust asks me to give a fully qualified path to specify the expected types, and the result is much uglier.

[the remark this answers is not in the document]
Comment (chriseth) [reply parser/src/powdr.lalrpop:453]: What about writing a parser helper function?
```
fn to_expr_with_source_ref<T: Into<Expression>>(inner_expr: T, offset: usize) -> Box<Expression> {
  let mut e = Expression::from(inner_expr);
  e.set_source_ref(offset);
  Box::new(e)
}
```
Also if this is the only use of `set_source_ref`, I would prefer to inline its current definition, because we should not need to set the source ref that often, I think...

File: parser/src/lib.rs [hunk 5]
```diff
@@ -43,6 +44,17 @@ impl ParserContext {
...
     }
+
+    pub fn to_expr_with_source_ref<T: Into<Expression>>(
+        &self,
+        inner_expr: T,
+        offset: usize,
+    ) -> Box<Expression> {
+        let mut expr = inner_expr.into();
+        let sr = self.source_ref(offset);
+        *expr.source_reference_mut() = sr;
```
Comment (chriseth) [inline parser/src/lib.rs:55]: Why the local variable?

File: ast/src/parsed/display.rs [hunk 6]
```diff
@@ -598,28 +598,37 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
...
+                writeln!(f, "match {scrutinee} {{")?;
+                write_items_indented(f, arms)?;
+                write!(f, "}}")
             }
-            Expression::IfExpression(e) => write!(f, "{e}"),
-            Expression::BlockExpression(block_expr) => {
-                write!(f, "{block_expr}")
+            Expression::IfExpression(_, e) => write!(f, "{e}"),
+            Expression::BlockExpression(_, BlockExpression { statements, expr }) => {
+                if statements.is_empty() {
```
Comment (chriseth) [inline ast/src/parsed/display.rs:624]: Looks like an old version as well.

File: ast/src/parsed/mod.rs [hunk 7]
```diff
@@ -332,24 +332,97 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+    BinaryOperation(SourceRef, BinaryOperation<Self>),
+    IndexAccess(SourceRef, IndexAccess<Self>),
+    FunctionCall(SourceRef, FunctionCall<Self>),
+    FreeInput(SourceRef, Box<Self>),
+    MatchExpression(SourceRef, MatchExpression<Self>),
+    IfExpression(SourceRef, IfExpression<Self>),
+    BlockExpression(SourceRef, BlockExpression<Self>),
+}
+
+// Comparison function for expressions that ignore source information.
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:355]:
```suggestion
/// Comparison function for expressions that ignore source information.
```

File: ast/src/parsed/mod.rs [hunk 8]
```diff
@@ -332,24 +332,97 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+            (Expression::ArrayLiteral(_, a), Expression::ArrayLiteral(_, b)) => a == b,
+            (Expression::BinaryOperation(_, a), Expression::BinaryOperation(_, b)) => a == b,
+            (Expression::UnaryOperation(_, a), Expression::UnaryOperation(_, b)) => a == b,
+            (Expression::IndexAccess(_, a), Expression::IndexAccess(_, b)) => a == b,
+            (Expression::FunctionCall(_, a), Expression::FunctionCall(_, b)) => a == b,
+            (Expression::FreeInput(_, a), Expression::FreeInput(_, b)) => a == b,
+            (Expression::MatchExpression(_, a), Expression::MatchExpression(_, b)) => a == b,
+            (Expression::IfExpression(_, a), Expression::IfExpression(_, b)) => a == b,
+            (Expression::BlockExpression(_, a), Expression::BlockExpression(_, b)) => a == b,
+            _ => false,
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:374]: I think this could be quite dangerous. If we add something to Expression, we will probably not extend this implementation.

Is there a different way of doing this? We probably don't want to clone and set the source ref, but maybe some simple macro?

File: ast/src/parsed/mod.rs [hunk 9]
```diff
@@ -332,24 +332,97 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+pub trait SourceReference {
+    fn source_reference(&self) -> &SourceRef;
+    fn source_reference_mut(&mut self) -> &mut SourceRef;
+}
+
+impl<E> SourceReference for Expression<E> {
+    fn source_reference(&self) -> &SourceRef {
+        use Expression::*;
+        match self {
+            Reference(src, _) => src,
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:388]: Maybe a macro would be useful here as well :)

File: ast/src/parsed/mod.rs [hunk 10]
```diff
@@ -442,34 +512,25 @@ pub struct MatchExpression<E = Expression<NamespacedPolynomialReference>> {
 
 impl<Ref> From<MatchExpression<Expression<Ref>>> for Expression<Ref> {
     fn from(match_expr: MatchExpression<Expression<Ref>>) -> Self {
-        Expression::MatchExpression(match_expr)
+        Expression::MatchExpression(SourceRef::unknown(), match_expr)
     }
 }
 
-impl<E> Children<E> for MatchExpression<E> {
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:449]: This might also be a merge conflict resolution error.

File: ast/src/parsed/mod.rs [hunk 11]
```diff
@@ -520,32 +581,40 @@ impl<R> Expression<R> {
...
-            Expression::Tuple(v) => v.iter(),
-            Expression::LambdaExpression(lambda) => lambda.children(),
-            Expression::ArrayLiteral(ArrayLiteral { items }) => items.iter(),
-            Expression::BinaryOperation(BinaryOperation { left, right, .. }) => {
+            Expression::Reference(_, _)
+            | Expression::PublicReference(_, _)
+            | Expression::String(_, _) => empty(),
+            Expression::Number(_, _) => empty(),
+            Expression::Tuple(_, v) => v.iter(),
+            Expression::LambdaExpression(_, LambdaExpression { body, .. }) => once(body.as_ref()),
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:589]: This should call children on the inner lambda expression

File: ast/src/parsed/mod.rs [hunk 12]
```diff
@@ -559,32 +628,40 @@ impl<R> Expression<R> {
...
-            Expression::Tuple(v) => v.iter_mut(),
-            Expression::LambdaExpression(lambda) => lambda.children_mut(),
-            Expression::ArrayLiteral(ArrayLiteral { items }) => items.iter_mut(),
-            Expression::BinaryOperation(BinaryOperation { left, right, .. }) => {
+            Expression::Reference(_, _)
+            | Expression::PublicReference(_, _)
+            | Expression::String(_, _) => empty(),
+            Expression::Number(_, _) => empty(),
+            Expression::Tuple(_, v) => v.iter_mut(),
+            Expression::LambdaExpression(_, LambdaExpression { body, .. }) => once(body.as_mut()),
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:636]: Same here, also for the others.

File: ast/src/parsed/display.rs (hunk 6 above)
Comment (gzanitti) [reply ast/src/parsed/display.rs:624]: Oh, sorry, I must have forgotten some branch of the match during the merge, how embarrassing. It won't happen again :sweat:

File: linker/src/lib.rs [hunk 13]
```diff
@@ -230,13 +230,10 @@ mod test {
     };
     use powdr_number::{BigUint, FieldElement, GoldilocksField};
 
+    use crate::{link, DEFAULT_DEGREE};
```
Comment (chriseth) [inline linker/src/lib.rs:233]: The changes in this file seem unrelated.

File: parser/src/test_utils.rs [hunk 14]
```diff
@@ -20,6 +24,30 @@ pub fn pil_statement_clear_source_ref(stmt: &mut PilStatement) {
         | PilStatement::EnumDeclaration(s, _) => *s = SourceRef::unknown(),
     }
 }
+
+fn pil_expression_clear_source_ref(expr: &mut powdr_ast::parsed::Expression) {
+    use powdr_ast::parsed::Expression::*;
+    match expr {
```
Comment (chriseth) [inline parser/src/test_utils.rs:30]: Can't you call `source_ref_mut()` here?

File: ast/src/parsed/mod.rs (hunk 8 above)
Comment (gzanitti) [reply ast/src/parsed/mod.rs]: I ended up implementing all the comparison inside a macro and using all the variants to make the match exhaustive (without using wildcards), in case we add a new variant to Expressions, the match will complain.

What do you think?

File: ast/src/parsed/mod.rs [hunk 15]
```diff
@@ -353,29 +353,42 @@ pub enum Expression<Ref = NamespacedPolynomialReference> {
...
-            _ => false,
+macro_rules! impl_partial_eq_for_expression {
+    ($($variant:ident),*) => {
+        impl<Ref: PartialEq> PartialEq for Expression<Ref> {
+            fn eq(&self, other: &Self) -> bool {
+                match (self, other) {
+                    $(
+                        (Expression::$variant(_, a), Expression::$variant(_, b)) => a == b,
+                    )*
+                    // This catches the case where variants are different and returns false
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:364]: Nice!

File: ast/src/parsed/mod.rs [hunk 16]
```diff
@@ -353,29 +353,42 @@ pub enum Expression<Ref = NamespacedPolynomialReference> {
...
+                    $(
+                        (Expression::$variant(_, _), _) => false,
+                    )*
+                }
+            }
         }
     }
 }
 
+impl_partial_eq_for_expression!(
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:374]: I think the macro I mentioned could be used here as well

**Review by chriseth — APPROVED** [review]
